File: src/types/query.ts
export type CellValue = string | number | boolean | null | { type: 'binary'; length: number };

export interface ColumnDef {
  name: string;
  dataType: string;
  nullable: boolean;
  ordinal: number;
}

export interface ResultSet {
  columns: ColumnDef[];
  rows: CellValue[][];
  rowCount: number;
  totalCount?: number;
  isTruncated?: boolean;
}

export interface QueryMessage {
  level: 'info' | 'warning' | 'error';
  message: string;
  code?: number;
  lineNumber?: number;
  timestamp: string;
  id?: string;
  seq?: number;
}

export interface QueryResult {
  resultSets: ResultSet[];
  messages: QueryMessage[];
  affectedRows: number;
  executionTimeMs: number;
}

export interface QueryHistoryItem {
  id: string;
  connectionId: string;
  database: string;
  sql: string;
  executedAt: string;
  executionTimeMs: number;
  status: 'success' | 'error' | 'cancelled';
  affectedRows?: number;
  errorMessage?: string;
  seq?: number;
}

export interface SessionMessageItem {
  id: string;
  seq: number;
  level: 'info' | 'warning' | 'error';
  message: string;
  code?: number;
  lineNumber?: number;
  timestamp: string;
}

export interface QueryResultTab {
  id: string;
  title: string;
  sql: string;
  result: QueryResult;
  executedAt: string;
  isPinned: boolean;
  durationMs: number;
  rowCount: number;
  connectionId?: string;
  database?: string;
  tableName?: string;
  schema?: string;
  seq?: number;
  isShowplan?: boolean;
}

File: src-tauri/src/models/query.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum CellValue {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    String(String),
    Binary {
        #[serde(rename = "type")]
        kind: String,
        length: usize,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnDef {
    pub name: String,
    pub data_type: String,
    pub nullable: bool,
    pub ordinal: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResultSet {
    pub columns: Vec<ColumnDef>,
    pub rows: Vec<Vec<CellValue>>,
    pub row_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryMessage {
    pub level: String, // 'info' | 'warning' | 'error'
    pub message: String,
    pub code: Option<i32>,
    pub line_number: Option<u32>,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryResult {
    pub result_sets: Vec<ResultSet>,
    pub messages: Vec<QueryMessage>,
    pub affected_rows: u64,
    pub execution_time_ms: u64,
}
